core/src/main/java/io/questdb/cutlass/pgwire/PGConnectionContext.java

```diff
@@ -2665,34 +2689,30 @@ private void setupFactoryAndCursor(SqlCompiler compiler) throws SqlException {
             boolean recompileStale = true;
             SqlExecutionCircuitBreaker circuitBreaker = sqlExecutionContext.getCircuitBreaker();
 
-            try {
-                if (!circuitBreaker.isTimerSet()) {
-                    circuitBreaker.resetTimer();
-                }
+            if (!circuitBreaker.isTimerSet()) {
+                circuitBreaker.resetTimer();
+            }
 
-                for (int retries = 0; recompileStale; retries++) {
-                    currentFactory = typesAndSelect.getFactory();
-                    try {
-                        currentCursor = currentFactory.getCursor(sqlExecutionContext);
-                        recompileStale = false;
-                        // cache random if it was replaced
-                        this.rnd = sqlExecutionContext.getRandom();
-                    } catch (TableReferenceOutOfDateException e) {
-                        if (retries == TableReferenceOutOfDateException.MAX_RETRY_ATTEMPS) {
-                            throw e;
-                        }
-                        LOG.info().$(e.getFlyweightMessage()).$();
-                        freeFactory();
-                        compileQuery(compiler);
-                        buildSelectColumnTypes();
-                        applyLatestBindColumnFormats();
-                    } catch (Throwable e) {
-                        freeFactory();
+            for (int retries = 0; recompileStale; retries++) {
+                currentFactory = typesAndSelect.getFactory();
+                try {
+                    currentCursor = currentFactory.getCursor(sqlExecutionContext);
+                    recompileStale = false;
+                    // cache random if it was replaced
+                    this.rnd = sqlExecutionContext.getRandom();
+                } catch (TableReferenceOutOfDateException e) {
+                    if (retries == TableReferenceOutOfDateException.MAX_RETRY_ATTEMPS) {
                         throw e;
                     }
+                    LOG.info().$(e.getFlyweightMessage()).$();
+                    freeFactory();
+                    compileQuery(compiler);
+                    buildSelectColumnTypes();
+                    applyLatestBindColumnFormats();
+                } catch (Throwable e) {
+                    freeFactory();
+                    throw e;
                 }
-            } finally {
-                circuitBreaker.unsetTimer();
```

Comment:
Why this try / finally is not necessary?

Reply:
That's because Andrei moved things like building hash map from getCursor() to first cursor.next() calls.